homeassistant/components/withings/__init__.py

```diff
@@ -126,27 +129,65 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
             hass, entry
         ),
     )
+    coordinator = WithingsDataUpdateCoordinator(hass, client)
 
-    use_webhooks = entry.options[CONF_USE_WEBHOOK]
-    coordinator = WithingsDataUpdateCoordinator(hass, client, use_webhooks)
-    if use_webhooks:
+    await coordinator.async_config_entry_first_refresh()
 
-        @callback
-        def async_call_later_callback(now) -> None:
-            hass.async_create_task(coordinator.async_subscribe_webhooks())
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
+
+    async def unregister_webhook(
+        call_or_event_or_dt: ServiceCall | Event | datetime | None,
+    ) -> None:
+        LOGGER.debug("Unregister Withings webhook (%s)", entry.data[CONF_WEBHOOK_ID])
+        webhook_unregister(hass, entry.data[CONF_WEBHOOK_ID])
+        await hass.data[DOMAIN][entry.entry_id].async_unsubscribe_webhooks()
+
+    async def register_webhook(
+        call_or_event_or_dt: ServiceCall | Event | datetime | None,
+    ) -> None:
+        if cloud.async_active_subscription(hass):
+            webhook_url = await async_cloudhook_generate_url(hass, entry)
+        else:
+            webhook_url = webhook_generate_url(hass, entry.data[CONF_WEBHOOK_ID])
+
+        if not webhook_url.startswith("https://"):
+            LOGGER.warning(
+                "Webhook not registered - "
+                "https and port 443 is required to register the webhook"
+            )
+            return
 
-        entry.async_on_unload(async_call_later(hass, 1, async_call_later_callback))
-        webhook.async_register(
+        webhook_register(
             hass,
             DOMAIN,
-            "Withings notify",
+            "Withings",
             entry.data[CONF_WEBHOOK_ID],
             get_webhook_handler(coordinator),
         )
 
-    await coordinator.async_config_entry_first_refresh()
+        await hass.data[DOMAIN][entry.entry_id].async_subscribe_webhooks(webhook_url)
+        LOGGER.debug("Register Withings webhook: %s", webhook_url)
+        entry.async_on_unload(
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, unregister_webhook)
+        )
 
-    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
+    async def manage_cloudhook(state: cloud.CloudConnectionState) -> None:
+        if state is cloud.CloudConnectionState.CLOUD_CONNECTED:
+            await register_webhook(None)
+
+        if state is cloud.CloudConnectionState.CLOUD_DISCONNECTED:
+            await unregister_webhook(None)
+            async_call_later(hass, 30, register_webhook)
+
+    if cloud.async_active_subscription(hass):
+        if cloud.async_is_connected(hass):
+            await register_webhook(None)
+        cloud.async_listen_connection_change(hass, manage_cloudhook)
+
+    elif hass.state == CoreState.running:
```

Comment:
Probably use this helper instead:
https://github.com/home-assistant/core/blob/65f307fe9c7f3cc2ffab063cef2f162b71da9856/homeassistant/helpers/start.py#L73